deeplake/util/transform.py

```diff
@@ -122,53 +104,141 @@ def store_data_slice(transform_input: Tuple) -> Dict:
-        data_slice = add_cache_to_dataset_slice(data_slice, tensors)
 
-    if extend_only:
-        extend_data_slice(
-            data_slice,
-            offset,
-            pipeline,
-            all_chunk_engines,
-            group_index,
-            pg_callback,
-        )
-    else:
-        transform_data_slice_and_append(
-            data_slice,
-            offset,
-            pipeline,
-            visible_tensors,
-            label_temp_tensors,
-            actual_tensors,
-            all_chunk_engines,
-            group_index,
-            pg_callback,
-            skip_ok,
-            ignore_errors,
+def _check_pipeline(out, tensors, skip_ok):
+    data = out.data
+    result_keys = set(k for k in data if not data[k].is_group and len(data[k]) > 0)
+
+    if skip_ok:
+        if not result_keys.issubset(tensors):
+            raise TensorMismatchError(list(tensors), list(result_keys), skip_ok)
+    elif set(result_keys) != set(tensors):
+        raise TensorMismatchError(list(tensors), list(result_keys), skip_ok)
+
+
+def write_sample_to_transform_dataset(out, transform_dataset):
+    if not is_empty_transform_dataset(out):
+        for tensor in out.tensors:
+            out_tensor = out[tensor]
+            transform_tensor = transform_dataset[tensor]
+            if transform_tensor.numpy_only and out_tensor.numpy_only:
+                for item in out_tensor.items:
+                    transform_tensor.extend(item)
+            else:
+                out_tensor.non_numpy_only()
+                transform_tensor.extend(out_tensor.items)
+            out_tensor.items.clear()
+
+
+def _handle_transform_error(
+    data_slice,
+    offset,
+    transform_dataset,
+    pipeline,
+    tensors,
+    end_input_idx,
+    ignore_errors,
+):
+    start_input_idx = transform_dataset.start_input_idx
+    for i in range(start_input_idx, end_input_idx + 1):
+        sample = (
+            data_slice[i : i + 1]
+            if pd and isinstance(data_slice, pd.DataFrame)
+            else data_slice[i]
         )
+        try:
+            out = transform_sample(sample, pipeline, tensors)
 
-    # retrieve relevant objects from memory
+            write_sample_to_transform_dataset(out, transform_dataset)
+
+            transform_dataset.flush()
+        except Exception as e:
+            if isinstance(e, SampleAppendError) and ignore_errors:
+                continue
+            raise TransformError(offset + i, sample) from e
+
+
+def _transform_and_append_data_slice(
+    data_slice,
+    offset,
+    transform_dataset,
+    pipeline,
+    tensors,
+    skip_ok,
+    pg_callback,
+    ignore_errors,
+):
+    n = len(data_slice)
+
+    pipeline_checked = False
+
+    for i, sample in enumerate(
+        (data_slice[i : i + 1] for i in range(n))
+        if pd and isinstance(data_slice, pd.DataFrame)
+        else data_slice
+    ):
+        try:
+            transform_dataset.set_start_input_idx(i)
+
+            try:
```

Comment:
this try will catch error while a sample is being transformed. it is easier to handle compared to an error after a sample has been sent to the chunk engine.